src/main/java/htsjdk/samtools/reference/FastaReferenceWriter.java

```diff
@@ -0,0 +1,758 @@
+        // for the sake of avoiding creating output if basesPerLine is invalid.
+        this.defaultBasePerLine = checkBasesPerLine(basesPerLine);
+
+        this.fastaStream = new CountingOutputStream(Files.newOutputStream(fastaFile));
+        this.indexWriter = indexFile == null ? NullWriter.NULL_WRITER : new OutputStreamWriter(Files.newOutputStream(indexFile), CHARSET);
+        final BufferedWriter dictWriter = new BufferedWriter(dictFile == null ? NullWriter.NULL_WRITER : new OutputStreamWriter(Files.newOutputStream(dictFile), CHARSET));
+        this.dictWriter = dictWriter;
+        this.dictCodec = new SAMSequenceDictionaryCodec(dictWriter);
+        this.dictCodec.encodeHeaderLine(false);
+        this.sequenceNamesAndSizes = new LinkedHashMap<>();
+    }
+
+    /**
+     * Creates a reference FASTA file writer.
+     * <p>
+     * You can specify a specific output stream to each file: the main fasta output, its index and its dictionary.
+     * </p>
+     *
+     * @param fastaOutput the output fasta file path.
+     * @param indexOutput the output stream to the index file, if requested, {@code null} if none should be generated.
+     * @param dictOutput  the output stream to the dictFile, if requested, {@code null} if none should be generated.
+     * @throws IllegalArgumentException if {@code fastaFile} is {@code null} or {@code basesPerLine} is 0 or negative.
+     */
+    public FastaReferenceWriter(final OutputStream fastaOutput,
+                                final int basesPerLine,
+                                final OutputStream indexOutput,
+                                final OutputStream dictOutput) {
+        this.defaultBasePerLine = checkBasesPerLine(basesPerLine);
+        this.fastaStream = new CountingOutputStream(fastaOutput);
+        this.indexWriter = indexOutput == null ? NullWriter.NULL_WRITER : new OutputStreamWriter(indexOutput, CHARSET);
+        final BufferedWriter dictWriter = new BufferedWriter(dictOutput == null ? NullWriter.NULL_WRITER : new OutputStreamWriter(dictOutput, CHARSET));
+        this.dictWriter = dictWriter;
+        this.dictCodec = new SAMSequenceDictionaryCodec(dictWriter);
+        this.dictCodec.encodeHeaderLine(false);
+        this.sequenceNamesAndSizes = new LinkedHashMap<>();
+    }
+
+    private static Path defaultFaiFile(final boolean makeFaiFile, final Path fastaFile) {
+        return makeFaiFile ? ReferenceSequenceFileFactory.getFastaIndexFileName(fastaFile) : null;
+    }
+
+    private static Path defaultDictFile(final boolean makeDictFile, final Path fastaFile) {
+        return makeDictFile ? ReferenceSequenceFileFactory.getDefaultDictionaryForReferenceSequence(fastaFile) : null;
+    }
+
+    // checks that a sequence name is valid.
+    private static void checkSequenceName(final String name) {
+        ValidationUtils.nonNull(name ,"Sequence name should not be null");
+        ValidationUtils.nonEmpty(name,"Sequence name should not be empty");
```

Comment:
it seems unnecessary to force a caller to use both `nonNull()` and `nonEmpty()` in this case. As a caller, we'd want to assume that `nonEmpty()` will also check for null.

If you really want the user to get different messages for null vs empty, the usual way to solve this problem (validation of inputs with customized messages) is to pass in the name of the thing you're validating, so  the validation code can have more than one failure case.